File: solutions/113.Path_Sum_II/AC_dfs_n.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Author:      illuz <iilluzen[at]gmail.com>
# File:        AC_dfs_n.py
# Create Date: 2015-02-21 09:05:37
# Usage:       AC_dfs_n.py 
# Descripton:  


# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def dfs(self, root, sum):
        if root == None:
            return
        self.path.append(root.val)
        sum -= root.val
        if root.left == None and root.right == None:
            if sum == 0:
                self.res.append(self.path[:])
        else:
            if root.left:
                self.dfs(root.left, sum)
            if root.right:
                self.dfs(root.right, sum)
        self.path.pop()

    # @param root, a tree node
    # @param sum, an integer
    # @return a list of lists of integers
    def pathSum(self, root, sum):
        self.res = []
        self.path = []
        self.dfs(root, sum)
        return self.res


# debug
root = TreeNode(1)
sum = 1
s  = Solution()
print s.pathSum(root, sum)

File: solutions/113.Path_Sum_II/AC_dfs_n.java
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_dfs_n.java
*  Create Date: 2015-02-21 09:23:19
*  Descripton:   
*/

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Solution {

    // Definition for binary tree
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }

    private void dfs(TreeNode root, int sum, List<List<Integer>> res, List<Integer> path) {
        if (root == null)
            return;
        path.add(root.val);
        sum -= root.val;
        if (root.left == null && root.right == null && sum == 0) {
            res.add(new ArrayList<Integer>(path));
        }
        dfs(root.left, sum, res, path);
        dfs(root.right, sum, res, path);
        path.remove(path.size() - 1);
    }

    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> path = new ArrayList<Integer>();
        dfs(root, sum, res, path);
        return res;
    }

    // debug
    public static void main(String[] args) {
        Scanner cin = new Scanner(System.in);
        Solution s = new Solution();
        int[] input = {1, 2, 3, 1};
        System.out.println("no case");
    }
}
